File: DataLoader.py
import requests
import zlib
import time


class Item:
    def __init__(self, open_, close, high, low, volume, amount):
        self.open = open_
        self.close = close
        self.high = high
        self.low = low
        self.volume = volume
        self.amount = amount

    def __repr__(self):
        return (f"Item(open={self.open}, close={self.close}, high={self.high}, "
                f"low={self.low}, volume={self.volume}, amount={self.amount})")


def decrypt(encrypted_bytes: bytes) -> bytes:
    predefined_xor = b"ihq_exp_hello"
    decrypted = bytearray(len(encrypted_bytes))
    for i, b in enumerate(encrypted_bytes):
        b = (~b) & 0xFF
        decrypted[i] = b ^ predefined_xor[i % len(predefined_xor)]
    return bytes(decrypted)


def get_if_legend(part: str) -> str:
    if len(part) != 12:
        return None
    a, b = part[:6], part[6:]
    if a != b:
        raise IOError(f"Invalid legend part: {part}")
    return a


def main():
    start = time.time()
    url = "https://gh-proxy.com/https://github.com/huzpsb/ihq_bin/raw/refs/heads/main/ihq.bin"
    file_name = "ihq.bin"
    print("Downloading ihq.bin ...")
    with requests.get(url, stream=True) as r:
        r.raise_for_status()
        with open(file_name, "wb") as f:
            for chunk in r.iter_content(chunk_size=8192):
                f.write(chunk)
    with open(file_name, "rb") as f:
        encrypted_bytes = f.read()
    decrypted_bytes = decrypt(encrypted_bytes)
    decrypted_bytes = zlib.decompress(decrypted_bytes)
    content = decrypted_bytes.decode("utf-8")
    lines = content.splitlines()
    legends = lines[0].split(",")
    assert len(legends) == 31
    all_items = []
    codes = []
    legend_now = None
    items_now = []
    for line in lines[1:]:
        parts = line.split(",")
        if len(parts) == 1:
            if parts[0] == "EOF":
                break
            elif parts[0] == "E":
                if legend_now is not None and len(items_now) == 31:
                    all_items.append(items_now)
                    codes.append(legend_now)
                    items_now = []
                    legend_now = None
                else:
                    raise IOError("Unexpected E")
            else:
                raise IOError(f"Unexpected line: {parts[0]}")
        elif len(parts) == 7:
            if legend_now is None:
                legend_now = get_if_legend(parts[0])
                if legend_now is None:
                    raise IOError(f"Expected legend, got: {parts[0]}")
            else:
                if legend_now != parts[0][:6]:
                    raise IOError(f"Legend mismatch: {legend_now} vs {parts[0]}")
            open_, close, high, low, volume, amount = map(float, parts[1:])
            items_now.append(Item(open_, close, high, low, volume, amount))
        else:
            raise IOError(f"Unexpected line length: {len(parts)}")

    items = all_items
    code_to_index = {code: i for i, code in enumerate(codes)}
    date_to_index = {date: i for i, date in enumerate(legends)}
    print(f"Data loaded in {int((time.time() - start) * 1000)} ms.")
    print(f"Loaded {len(codes)} codes.")
    print(f"Dates: {legends}")
    example_code = codes[0]
    example_date = legends[0]
    code_index = code_to_index[example_code]
    date_index = date_to_index[example_date]
    example_item = items[code_index][date_index]
    print(f"Example item for code {example_code} on date {example_date}:")
    print(example_item)


if __name__ == "__main__":
    main()

File: DataLoader.java
import java.io.*;
import java.net.HttpURLConnection;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.zip.InflaterInputStream;

public class DataLoader {
    private static byte[] decrypt(byte[] encryptedBytes) {
        byte[] decrypted = new byte[encryptedBytes.length];
        byte[] predefinedXor = "ihq_exp_hello".getBytes(StandardCharsets.UTF_8);
        for (int i = 0; i < encryptedBytes.length; i++) {
            byte b = (byte) (~encryptedBytes[i]);
            decrypted[i] = (byte) (b ^ predefinedXor[i % predefinedXor.length]);
        }
        return decrypted;
    }

    public static void main(String[] args) throws Exception {
        long now = System.currentTimeMillis();
        FileOutputStream fos = new FileOutputStream("ihq.bin");
        HttpURLConnection conn = (HttpURLConnection) new java.net.URL("https://gh-proxy.com/https://github.com/huzpsb/ihq_bin/raw/refs/heads/main/ihq.bin").openConnection();
        conn.setRequestMethod("GET");
        InputStream is = conn.getInputStream();
        byte[] buffer = new byte[8192];
        int bytesRead;
        while ((bytesRead = is.read(buffer)) != -1) {
            fos.write(buffer, 0, bytesRead);
        }
        is.close();
        fos.close();


        FileInputStream fis = new FileInputStream("ihq.bin");
        byte[] encryptedBytes = fis.readAllBytes();
        byte[] decryptedBytes = decrypt(encryptedBytes);
        fis.close();
        ByteArrayInputStream bais = new ByteArrayInputStream(decryptedBytes);
        InflaterInputStream iis = new InflaterInputStream(bais);
        decryptedBytes = iis.readAllBytes();
        iis.close();
        String content = new String(decryptedBytes, StandardCharsets.UTF_8);

        Scanner scanner = new Scanner(content);
        String[] legends = scanner.nextLine().split(",");
        assert legends.length == 31;

        String legendNow = null;
        List<Item> itemsNow = new ArrayList<>();
        List<String> codes = new ArrayList<>();
        List<List<Item>> allItems = new ArrayList<>();
        s:
        while (true) {
            if (!scanner.hasNextLine()) throw new IOException("Unexpected end of file");
            String[] line = scanner.nextLine().split(",");
            if (line.length == 1) {
                switch (line[0]) {
                    case "EOF":
                        break s;
                    case "E":
                        if (legendNow != null && itemsNow.size() == 31) {
                            allItems.add(itemsNow);
                            codes.add(legendNow);
                            itemsNow = new ArrayList<>();
                            legendNow = null;
                            continue;
                        } else {
                            throw new IOException("Unexpected E");
                        }
                    default:
                        throw new IOException("Unexpected line: " + line[0]);
                }
            }
            if (line.length == 7) {
                if (legendNow == null) {
                    legendNow = getIfLegend(line[0]);
                    if (legendNow == null) {
                        throw new IOException("Expected legend, got: " + line[0]);
                    }
                } else {
                    if (!legendNow.equals(line[0])) {
                        throw new IOException("Legend mismatch: " + legendNow + " vs " + line[0]);
                    }
                }
                double open = Double.parseDouble(line[1]);
                double close = Double.parseDouble(line[2]);
                double high = Double.parseDouble(line[3]);
                double low = Double.parseDouble(line[4]);
                double volume = Double.parseDouble(line[5]);
                double amount = Double.parseDouble(line[6]);
                itemsNow.add(new Item(open, close, high, low, volume, amount));
                continue;
            }
            throw new IOException("Unexpected line length: " + line.length);
        }
        scanner.close();
        Item[][] items = new Item[codes.size()][31];
        for (int i = 0; i < codes.size(); i++) {
            items[i] = allItems.get(i).toArray(new Item[0]);
        }
        Map<String, Integer> codeToIndex = new HashMap<>();
        for (int i = 0; i < codes.size(); i++) {
            codeToIndex.put(codes.get(i), i);
        }
        Map<String, Integer> dateToIndex = new HashMap<>();
        for (int i = 0; i < legends.length; i++) {
            dateToIndex.put(legends[i], i);
        }
        System.out.println("Data loaded in " + (System.currentTimeMillis() - now) + " ms.");

        // Example usage:
        System.out.println("Loaded " + codes.size() + " codes.");
        System.out.println("Dates: " + Arrays.toString(legends));
        String exampleCode = codes.get(0);
        String exampleDate = legends[0];
        int codeIndex = codeToIndex.get(exampleCode);
        int dateIndex = dateToIndex.get(exampleDate);
        Item exampleItem = items[codeIndex][dateIndex];
        System.out.println("Example item for code " + exampleCode + " on date " + exampleDate + ":");
        System.out.println(exampleItem);
    }

    private static String getIfLegend(String part) throws IOException {
        if (part.length() != 12) {
            return null;
        }
        String a = part.substring(0, 6);
        String b = part.substring(6, 12);
        if (!a.equals(b)) {
            throw new IOException("Invalid legend part: " + part);
        }
        return a;
    }
}

class Item {
    final double open;
    final double close;
    final double high;
    final double low;
    final double volume;
    final double amount;

    Item(double open, double close, double high, double low, double volume, double amount) {
        this.open = open;
        this.close = close;
        this.high = high;
        this.low = low;
        this.volume = volume;
        this.amount = amount;
    }

    @Override
    public String toString() {
        return "Item{" +
                "open=" + open +
                ", close=" + close +
                ", high=" + high +
                ", low=" + low +
                ", volume=" + volume +
                ", amount=" + amount +
                '}';
    }
}
